proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,500 @@
+  # unless they belong to one of these GitHub orgs.
+  # The GitHub App or OAuth App provided in the
+  # spec.client.secretName must be allowed by the org
+  # owners to view org memberships, or else the
+  # user will not be considered to belong to that org
+  # or belong to any teams within that org.
+  # When specified, users' group memberships will be
+  # filtered to include only those GitHub teams which
+  # are owned by these orgs.
+  # When not set, any GitHub user can authenticate
+  # regardless of org membership, and all the team
+  # memberships returned by the GitHub API for that
+  # user will be reflected as groups regardless of
+  # which orgs own those teams.
+  allowOrganizations: [ vmware-tanzu, broadcom ]
+  # Allow or disallow users to use GitHub Personal Access Tokens
+  # for non-interactive (no web browser) authentication.
+  # Setting these both to false disables all non-interactive
+  # CLI-based auth, but still allows browser-based auth.
+  allowPersonalAccessTokens:
+    # Allow or disallow the use of GitHub Fine-grained PATs
+    # to authenticate to Kubernetes clusters. If they
+    # are disallowed for your GitHub org, then also
+    # disallow them here so your users can get a nice
+    # error message when trying to use a fine-grained PAT.
+    # Defaults to false.
+    fineGrained: true
+    # Allow or disallow the use of GitHub Classic PATs
+    # to authenticate to Kubernetes clusters. If they
+    # are disallowed for your GitHub org, then also
+    # disallow them here so your users can get a nice
+    # error message when trying to use a classic PAT.
+    # Defaults to false.
+    classic: false
+  client:
+    # The name of Secret in the same namespace which holds the
+    # client ID and client secret of a GitHub App or OAuth App.
+    # This will only be used for web browser-based auth flows.
+    # Required.
+    secretName: github-client-credentials
+status:
+  conditions: # conditions TBD. Shows validations, if any.
+  phase: # ready or error, if there are conditions.
+```
+
+#### Web Browser-based Authentication using GitHubIdentityProvider
+
+The client ID and client secret configured on the GitHubIdentityProvider can be for either a GitHub App or an OAuth App.
+This flexibility can be made possible because there are many similarities in the OAuth 2.0 authorization flow used by
+both. To the extent that we can ignore the differences, Pinniped will not need to know if it is acting as a GitHub App
+or an OAuth App.
+
+The authorization flow for GitHub apps is described in
+[this doc](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-user-access-token-for-a-github-app)
+and the flow for OAuth Apps is described in
+[this doc](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps).
+
+From a client's point of view, the main difference is that access tokens for a GitHub App may expire, depending how the
```

Comment:
IE, a session is 8 or 9 hrs, depending on if it is a github session or a pinniped session.  Tokens bound to a session are short lived.  In the case of Github, the token is relatively long lived, for the entire life of the 8 hour session, then expire.  For other tokens in our current implementations, they are short lived, but the refresh flow is used to perpetually grant new tokens until the 9 hour session lifetime is reached, and then the user must login again.

Reply:
Yes.